vpr/src/analytical_place/ap_netlist.cpp

```diff
@@ -58,8 +58,11 @@ APBlockId APNetlist::create_block(const std::string& name, const t_pack_molecule
 void APNetlist::set_block_loc(const APBlockId id, const APFixedBlockLoc& loc) {
     VTR_ASSERT_SAFE(valid_block_id(id));
 
-    // Check that the location is fixed, if all values are -1 then it is not fixed.
-    if (loc.x == -1 && loc.y == -1 && loc.sub_tile == -1 && loc.layer_num == -1)
+    // Check that the location is fixed, if all dims are unfixed then it is not fixed.
+    if (loc.x == APFixedBlockLoc::UNFIXED_DIM &&
+        loc.y == APFixedBlockLoc::UNFIXED_DIM &&
+        loc.sub_tile == APFixedBlockLoc::UNFIXED_DIM &&
+        loc.layer_num == APFixedBlockLoc::UNFIXED_DIM)
         return;
 
     block_locs_[id] = loc;
```

Comment:
Seems a bit odd to allow fixed locations with only some dimensions fixed (not sure how you're going to handle that; it would seem like a strange/limited form of region constraint that isn't worth supporting since we already have more general region support).  I suggest you assert that all dimensions are fixed if any are fixed; code can then assume FIXED means truly locked down.  You can generalize that later, but I'd do that by going through the partition/region constraints, not by allowing individual dimensions to be -1.

Reply:
It actually is sufficient to just lock it down in the x-dimension, for example, since solvers tend to solve x independent to y. The layer_num is currently always assumed to be 0; but it was easy to add to this struct in this case (being forward looking).

The sub_tile actually need not be locked down (nor should it). The placement constraints file leaves this optional. This was the main reason UNFIXED_DIM exists.

However, since the generation of the APNetlist from the atoms never allows x, y, and layer to be unspecified, I have added an assert to this method to guard against them being unset. I have purposefully left sub_tile out of this assert.